### ISSUE TITLE
Older app version triggering update(fdroid)

### ISSUE THREAD

**Issue by Samandamand** [ImranR98/Obtainium#920]
Obtainium will alert me I have updates for apps. I batch update and almost all of them fail because they are older versions than what is already installed. This seems to only be happening with apps from Fdroid. I currently have 13 apps showing they have updates which are older than I have installed.

Examples:
[https://drive.proton.me/urls/4KBRXGZWBC#bUxq-b4d6O_q](url)
[https://drive.proton.me/urls/DVA57VZT24#Qoe4v-LGRm1_](url)
[https://drive.proton.me/urls/TMJ474QVF4#YoAwbkhaCgn5](url)

 - Device: Pixel 5a
 - OS: GrapheneOS
 - Obtainium Version: 0.14.17

**Comment by GenericMale** [comment]
F-Droid seems to have deleted a bunch of releases just now. Other examples:
Gadgetbridge 0.76.0 -> 0.75.1
Drinkable 1.37.0 -> 1.36.2

I don't think this should be a normal use case. Best thing obtainium could do is skip those downgrades and show a warning instead of the error popup.

**Comment by pled** [comment]
Yes, another app trigering updates when already installed with higher version : 
OsmAnd~ 
It is a bit strange because 4.5.10 is installed on the smartphone (some days ago), Obtainium prompts for update with same version : 
![obt1](https://github.com/ImranR98/Obtainium/assets/2879483/363fe7bc-9e1f-4be8-a77f-493b89229d96)
But looking at details, I see 4.5.8 as "last version" while on f-froid, last version is 4.5.10 ?
![obt2](https://github.com/ImranR98/Obtainium/assets/2879483/c94e7195-3a19-4a4d-9209-abd5cb1a8ff9)

**Comment by FAT9L** [comment]
Same here, with 4-5 of my F-Droid apps. 

![Screenshot_20230920_163237.png](https://github.com/ImranR98/Obtainium/assets/96025485/3c0783ea-1f38-4bb9-a6ab-ebca911a1305)

**Comment by ImranR98** [comment]
Could you try disabling "Select suggested versionCode APK" and see if that fixes it? It was added recently; maybe the F-Droid API isn't stable for that prop.

**Comment by FAT9L** [comment]
> Could you try disabling "Select suggested versionCode APK" and see if that fixes it? It was added recently; maybe the F-Droid API isn't stable for that prop.

No change for me - although I'm not the OP. I tried both "try selecting suggested versionCode APK" and "auto-select highest versionCode APK".

Disabling version detection (or setting an app to  track-only) is the only real workaround that I've found.

**Comment by pled** [comment]
> Could you try disabling "Select suggested versionCode APK" and see if that fixes it? It was added recently; maybe the F-Droid API isn't stable for that prop.

Well, this is only possible for new installation, so it does not solve the issue with current apps like OsmAnd in my case, and I don't want to uninstall it !

But I try with qBitController as a new app : having "Select suggested versionCode APK" selected or not makes no change after refresh, but no issue observed : version 8.1 is installed, as expected (this is the version available on f-droid) : 

![Screenshot_20230922-095532](https://github.com/ImranR98/Obtainium/assets/2879483/e28ec32c-ca44-4ec4-96dd-ec2a89923cac)


But I notice the 8.2 version is available on [github repository](https://github.com/Bartuzen/qBitController/releases/tag/v0.8.2) for that app ??

**Comment by ImranR98** [comment]
> But I notice the 0.8.2 version is available on [github repository](https://github.com/Bartuzen/qBitController/releases/tag/v0.8.2) for that app ?? so potential rollback on f-droid ?

There's no guarantee they will be in sync - F-Droid has their own app build process.

**Comment by pled** [comment]
Yes, I asked this for FAT9L screenshot.

**Comment by ImranR98** [comment]
Still haven't figured out what could be going on (or been able to reproduce the issue). In the case of Gadgetbridge and Drinkable, you @GenericMale mentioned that "F-Droid seems to have deleted a bunch of releases just now" - did you verify this was actually an issue on the F-Droid side?

**Comment by GenericMale** [comment]
Yeah i checked the latest version directly on the F-Droid site and it definitely was the older one.
In this case not an issue with Obtainium not picking up the latest one as far as i could tell.

I had Drinkable and Gadgetbridge installed with Obtainium via F-Droid since a long time and didn't change the configuration and never installed them manually from any other source. So a newer version must have been on F-Droid at one point before.
Obtainium alerted me of an "update" for the older version and after downloading gave me an error that an older version is already installed.

Obviously i can't reproduce it now because the latest versions are on F-Droid again.
I tried to check the f-droid forums, their issue tracker and reddit to see if there was a reason for the rollback of some releases but couldn't find anything.

**Comment by teemue** [comment]
Faced this problem after the latest app update. It's serious bug because it turns the app unusable.

**Comment by doug8888** [comment]
Same problem here this morning.

**Comment by DwainZwerg** [comment]
This problem is not unique to F-Droid: Whenever there is a newer version on the device (even if they have the same signature) than the one in the source given in Obtainium, Obtainium thinks there is an update, although it is in fact a downgrade. Actually, I have the problem that I have installed the AuroraStore (with my own Google account and activated beta updates) and sometimes get updates there that have not yet arrived in Apkpure and therefore not yet in Obtainium, so Obtainium suggests a downgrade.

**Comment by DeWarner** [comment]
I have had the same thing, in my case it was that the suggested version was the previous version. I fixed it by changing the settings for that app to ignore suggested version. Ideally the suggested version would be ignored depending on if the installed version is higher

**Comment by DwainZwerg** [comment]
But that's not the problem for me, because I have an even newer version than the package source even knows (I've already checked). So it doesn't matter which version I retrieve from the package source, it is always a version that is too old.

**Comment by ImranR98** [comment]
Some of the comments above seem to be about completely different things, unrelated to the original comment:

**1:**

> In my case it was that the suggested version was the previous version. I fixed it by changing the settings for that app to ignore suggested version.

"Use suggested version" was enabled by default because I thought F-Droid's suggestion would always be the highest version, not including beta or pre-releases. It has now been set to disabled by default so this shouldn't be an issue going forward (users who enable it manually will know what they're doing).

**2:**

> Because I have an even newer version than the package source even knows (I've already checked). So it doesn't matter which version I retrieve from the package source, it is always a version that is too old.

This is intended behaviour. Obtainium sees a difference between your currently installed version and the latest version from the source, and interprets this as an update. We don't actually compare the version numbers to see which is higher for various reasons discussed in other issues (might search and link here later). If your app is already being updated externally, it might not be useful to add it to Obtainium. Or you could switch off version detection and mark it as updated manually.

**3:**

> Yeah i checked the latest version directly on the F-Droid site and it definitely was the older one.

This was a problem with the F-Droid API - it seems like they added new versions of these apps and then removed them for a while. Hopefully this doesn't happen very often since it has only been reported by one user for a couple of apps. Not something Obtainium could do much about.

So just to clarify, for all 3 cases quoted above, the cause has either been fixed or is due to factors outside of the app. This issue is open to track the behaviour described by @Samandamand in their original comment (where 'QR Scanner', 'Antennapod', and 'Aurora Store' reported older "latest" versions). As far as I can tell it isn't explained by any of the 3 scenarios listed above (could have been due to case 3 - F-Droid API issue - but it's hard to know unless it happens again).

**Comment by pled** [comment]
And OsmAnd~ case as well, if I am correct.

**Comment by ImranR98** [comment]
> And OsmAnd~ case as well, if I am correct.

I saw that the "wrong version" for OsmAnd is the same as F-Droid's suggested version, so I assumed that it was like case 1. Looking through the comments, you didn't confirm if the toggle was off. Let me know if not.

**Comment by pled** [comment]
You are right. Disabling "Use suggested version" solved the issue. Initially, I was unable to find that option (I only looked at Optainium general parameters, not the "additional options" for an installed app. Sorry ! :-(

**Comment by DotDev262** [comment]
Now the problem is with GitHub as well for Libretube.

![Screenshot_20231119_000630.jpg](https://github.com/ImranR98/Obtainium/assets/127623817/1f3af895-dbfe-4f92-b188-5cf78ad9f715)

**Comment by jayb-g** [comment]
Indeed its inconvenient to have  downgrade notifications for whatever reasons whether its F-Droid or Aurora it does happen. Obtanium should be able to check if currently installed version is higher or lower than the proposed update version. I think that would fix the issue for all/most of above cases without causing further complications.

In case a user really wants to downgrade or has to downgrade, he has to uninstall the app anyway so it doesn't seem to be useful to have downgrade notifications, although might be ok just as a notice/warning.
